### PR TITLE
Add a client flag to delete "--now" for grace period 0

### PR DESCRIPTION
**Pull request by smarterclayton**
Is equivalent to --grace-period=0 but is more intuitive for end users.

Stuck pod?

```
kubectl delete pod foo --now
```

### CODE REVIEW

**Comment by ncdc** [comment]
Description should probably say kubectl instead of oc :smile:

**Comment by smarterclayton** [comment]
We should have called kubectl kc

**Comment by kargakis** [comment]
Can you also add an example?

cc: @kubernetes/kubectl

**Comment by smarterclayton** [comment]
@k8s-bot test this: #23878

**Comment by deads2k** [comment]
I don't think I want to use the `--force` flag for this.  I really want the idea of "protected" resources.  These would be resources that are hard to delete, but should still be deletable if someone really means it.  When we try to delete that resource, we could require them to add a "force" option to `DeleteOptions`.

As an example of things you want to protect, I could imagine someone deciding to protect serviceaccount/foo because they're using that one to run some maintenance script.  I'd say this is also the exact case that was "solved" with `immortalNamespaces`.

**Comment by smarterclayton** [comment]
Force is saying "delete with grace period 0 without waiting for it to
gracefully shutdown".  Nothing to do with immortal resources.  Unless
you're saying you want to preserve --force for use with deletion protected
resources, which I can buy.  In which case '--now' may be more appropriate.

The user expectation that needs to be managed is "I tried to delete this,
but it doesn't seem to be working.  I want to override the normal behavior
and force deletion.".  So force is probably aligned with the natural user
expectation.

**Comment by deads2k** [comment]
> Unless
> you're saying you want to preserve --force for use with deletion protected
> resources, which I can buy.

Correct.  Since the concept doesn't exist yet, I felt like I had to describe it.

I would prefer to have the concept of protected resources, in which case `--force` is a more natural fit for that.  I'm fine with `--now`.

**Comment by bgrant0607** [comment]
--force is about ignoring or working around conflicts.

**Comment by bgrant0607** [comment]
--now would be similar to `shutdown now`, so I'm in favor of that.

**Comment by bgrant0607** [comment]
cc @caesarxuchao @janetkuo

**Comment by smarterclayton** [comment]
Updated

**Comment by caesarxuchao** [comment]
@smarterclayton @bgrant0607 in the future, do you expect "--grace-period=0" and "--now" to mean different things?

#23656 is proposing adding finalizers to resources. In that context, I would imagine for the deletion of a pod, "--now" means `ignore all the finalizers, remove the Pod from etcd at once so that its name is released immediately`; and "--grace-period=0" means `containers receive a SIGKILL immediately, but Pod won't be deleted from etcd until all finalizers have completed`.

**Comment by smarterclayton** [comment]
Would removing finalizers be a thing regular users can do?

**Comment by caesarxuchao** [comment]
> Would removing finalizers be a thing regular users can do?

Yes, I think it's necessary. If a deletion is stuck because a finalizer is not responsive, a user should be able to remove the finalizer manually.

**Comment by bgrant0607** [comment]
@caesarxuchao Good point, but may need to work through some use cases to understand whether removing finalizers would be acceptable. I suggest we work that out in the cascading deletion proposal rather than here. For example, blanket finalizer removal could conflict with another flag to indicate whether children should be orphaned. 

Finalizers might need to be handled on a case-by-case basis.

**Comment by smarterclayton** [comment]
What if the finalizer is a link to an external system of record?  I
had assumed initializers are owned by the platform and not the end
user, but had not considered finalizers as user driven

File: pkg/kubectl/cmd/delete.go [hunk 1]
```diff
@@ -138,10 +142,18 @@ func RunDelete(f *cmdutil.Factory, out io.Writer, cmd *cobra.Command, args []str
 		}
 	}
 
+	gracePeriod := cmdutil.GetFlagInt(cmd, "grace-period")
+	if cmdutil.GetFlagBool(cmd, "now") {
+		if gracePeriod != -1 {
+			return fmt.Errorf("--now and --grace-period cannot be specified together")
```
Comment (janetkuo) [inline pkg/kubectl/cmd/delete.go]: we can allow `--now` and `--grace-period=0` to be set together (although it's redundant)

**Comment by janetkuo** [comment]
Please update the PR title and description to `--now`

**Comment by smarterclayton** [comment]
Do we want to encourage users to do that?  Why would a user ever set them
together?  We generally don't allow combinations that are equivalent to be
specified together (which is fairly consistent from a unix perspective)

**Comment by janetkuo** [comment]
@smarterclayton I agree.

**Comment by smarterclayton** [comment]
Any other comments on this?

**Comment by deads2k** [comment]
lgtm

**Comment by deads2k** [comment]
@smarterclayton how picky are we on release notes?  I labelled it that way just in case we want to note every CLI change, but let me know if that's not the case.

**Comment by smarterclayton** [comment]
It's a usability improvement so I think it's deserved.

**Pull request merged**